From 0119eac7edb607d612a95a69c8751f0a36e24b15 Mon Sep 17 00:00:00 2001
From: Dusty Mabe <dusty@dustymabe.com>
Date: Wed, 1 Jul 2026 22:08:16 -0400
Subject: [PATCH 5/9] util/makeefi: update find_efi_vendor_dir_name to accept
 source_tree

It will conveniently now accept either an efidir if the caller already
knows where one is or a source filesystem tree which can be passed to
find_efi_source_paths() to find the efidir from.

This fixes a bug (introduced a few commits back) in the
coreos.live-artifacts.mono stage where we had removed the vendor dir
from an efidir but then tried to find the vendor dir name on the same
efidir right after.
---
 osbuild/util/makeefi.py                       | 20 +++++++++++++++++--
 stages/org.osbuild.coreos.live-artifacts.mono | 14 +++++--------
 2 files changed, 23 insertions(+), 11 deletions(-)

diff --git a/osbuild/util/makeefi.py b/osbuild/util/makeefi.py
index fab56507..97d5e881 100644
--- a/osbuild/util/makeefi.py
+++ b/osbuild/util/makeefi.py
@@ -32,7 +32,23 @@ def find_efi_source_paths(tree):
     return [os.path.join(tree, 'usr/lib/bootupd/updates/EFI')]
 
 
-def find_efi_vendor_dir_name(efidir):
+def find_efi_vendor_dir_name(efidir=None, source_tree=None):
+    """
+    Searches either an EFI directory structure (efidir) or a source tree
+    (source_tree) for an EFI vendor directory name (usually fedora,
+    redhat, or centos). If source_tree is given the efidir will be found
+    first using find_efi_source_paths().
+    """
+    if source_tree and efidir:
+        raise ValueError("find_efi_vendor_dir_name: cannot pass both efidir and source_tree")
+    if not source_tree and not efidir:
+        raise ValueError("find_efi_vendor_dir_name: must pass either efidir or source_tree")
+    # If the user provided a source_tree and not an efidir then
+    # we must first find the efidir within the source_tree.
+    if source_tree:
+        # To find the string for the vendor ID we only need to inspect one
+        # of the source EFI directories. Grab one here.
+        efidir = find_efi_source_paths(source_tree)[0]
     # Find name of vendor directory for this distro. i.e. "fedora" or "redhat".
     dirs = [n for n in os.listdir(efidir) if n != "BOOT"]
     if len(dirs) != 1:
@@ -49,7 +65,7 @@ def mkefidir(efidir, source_tree):
     for path in find_efi_source_paths(source_tree):
         shutil.copytree(path, efidir, dirs_exist_ok=True)
 
-    vendor_id = find_efi_vendor_dir_name(efidir)
+    vendor_id = find_efi_vendor_dir_name(efidir=efidir)
 
     # Delete fallback and its CSV file.  Its purpose is to create
     # EFI boot variables, which we don't want when booting from
diff --git a/stages/org.osbuild.coreos.live-artifacts.mono b/stages/org.osbuild.coreos.live-artifacts.mono
index d136f708..2d9b7f7c 100755
--- a/stages/org.osbuild.coreos.live-artifacts.mono
+++ b/stages/org.osbuild.coreos.live-artifacts.mono
@@ -277,7 +277,7 @@ def gen_live_artifacts(paths, tree, filenames, deployed_tree, loop_client, versi
     test_fixture = check_for_test_fixture(deployed_tree)
     basearch = os.uname().machine
 
-    kargs_json = copy_configs_and_init_kargs_json(deployed_tree, paths, blsentry_kargs, volid)
+    kargs_json = copy_configs_and_init_kargs_json(deployed_tree, paths, blsentry_kargs, volid, basearch)
 
     # Add placeholder for Ignition CPIO file.  This allows an external tool,
     # `coreos-installer iso ignition embed`, to modify an existing ISO image
@@ -316,7 +316,7 @@ def gen_live_artifacts(paths, tree, filenames, deployed_tree, loop_client, versi
         # directory: efidir
         makeefi.mkefidir(efidir=efidir, source_tree=deployed_tree)
         # Discover the vendor ID (can be different on Fedora/RHEL/CentOS)
-        vendor_id = makeefi.find_efi_vendor_dir_name(efidir)
+        vendor_id = makeefi.find_efi_vendor_dir_name(source_tree=deployed_tree)
         # Get the grub configuration for booting from the ISO and write it out
         # to $efidir/BOOT/grub.cfg. See the comment in the function for explanation
         # of config used.
@@ -553,7 +553,7 @@ def mkrootfs_metal(paths, workdir, img_metal, fstype, fsoptions, loop_client):
 
 
 # pylint: disable=too-many-branches
-def copy_configs_and_init_kargs_json(deployed_tree, paths, blsentry_kargs, volid):
+def copy_configs_and_init_kargs_json(deployed_tree, paths, blsentry_kargs, volid, basearch):
     # Filter kernel arguments for substituting into ISO bootloader
     kargs_array = [karg for karg in blsentry_kargs
                    if karg.split('=')[0] not in LIVE_EXCLUDE_KARGS]
@@ -620,12 +620,8 @@ def copy_configs_and_init_kargs_json(deployed_tree, paths, blsentry_kargs, volid
     # [1] https://github.com/coreos/fedora-coreos-config/tree/testing-devel/live/EFI/fedora
     # [2] https://github.com/openshift/os/tree/master/live/EFI/vendor
     # [3] https://github.com/openshift/os/issues/954
-    #
-    # To find the string for the vendor ID we only need to inspect one
-    # of the source EFI directories. Grab one here.
-    efidir = makeefi.find_efi_source_paths(deployed_tree)[0]
-    if os.path.exists(efidir):
-        vendor_id = makeefi.find_efi_vendor_dir_name(efidir)
+    if basearch in ("x86_64", "aarch64"):
+        vendor_id = makeefi.find_efi_vendor_dir_name(source_tree=deployed_tree)
         grubfilepath = ensure_glob(os.path.join(paths["iso"], 'EFI/*/grub.cfg'), n=1)[0]
         srcdir = os.path.dirname(grubfilepath)
         dstdir = os.path.join(paths["iso"], f"EFI/{vendor_id}")
-- 
2.54.0

